odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1471 @@
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluated
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return ((1 / self._scaling) * self._orig_convex_conj_f.gradient *
+                (1 / self._scaling))
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the ConvexConjugateArgScaling
+        functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator
+
+        Returns
+        -------
+        out : Operator
+            Domain and range equal to domain of functional
+        """
+        return proximal_arg_scaling(self._orig_convex_conj_f.proximal,
+                                    scaling=(1 / self._scaling))(sigma)
+
+    # TODO: Should it be added?
+    # Note: This would only be valide when f is proper convex and lower-
+    # semincontinuous.
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the functional."""
+        raise NotImplementedError
+
+
+class ConvexConjugateLinearPerturb(Functional):
+    """ The ``Functional`` representing (F(.) + <y,.>)^*.
+
+    Calculate the convex conjugate functional perturbed function F(x) + <y,x>.
+
+    This is calculated according to the rule
+
+        (F(.) + <y,.>)^* (x) = F^*(x - y)
+
+    where ``y`` is the linear perturbation.
+
+    Parameters
+    ----------
+    convex_conj_f : `Functional`
+        Function corresponding to F^*.
+
+    y : Element in domain of F^*.
+
+    Notes
+    -----
+    For reference on the identity used, see [KP2015]_. Note that this is only
+    valide for functionals with a domain that is a Hilbert space.
+    """
+
+    def __init__(self, convex_conj_f, y):
+        """Initialize a ConvexConjugateLinearPerturb instance.
+
+        Parameters
+        ----------
+        convex_conj_f : `Functional`
+            Function corresponding to F^*.
+
+        y : Element in domain of F^*.
+        """
+        if y is not None and not isinstance(y, LinearSpaceVector):
+            raise TypeError('vector {!r} not None or a LinearSpaceVector'
+                            ' instance.'.format(y))
+
+        if y not in convex_conj_f.domain:
+            raise TypeError('vector {} not in the domain of the functional {}.'
+                            ''.format(y, convex_conj_f.domain))
+
+        super().__init__(domain=convex_conj_f.domain,
+                         linear=convex_conj_f.is_linear,
+                         smooth=convex_conj_f.is_smooth,
+                         concave=convex_conj_f.is_concave,
+                         convex=convex_conj_f.is_convex)
+
+        self._orig_convex_conj_f = convex_conj_f
+        self._y = y
```

Comment:
I have gone through the code ones. Note that there are TODO's left in the code; some of them contains open questions. I would greatly appreciate some input in those questions. Also: some of my review comments also contains questions; same thing applies here. :)